### PR TITLE
Fix parasite_axes does not properly handle units 

### PR DESCRIPTION
**Pull request by weiji-li**
## PR Summary

Right now parasite_axes just use self._parent_axes._get_lines as self._get_lines, but it can't update the axes unit when there are twin axes. Therefore, we need to provide axes=self as an extra args to handle this. We also need to change the callees to use axes in kwargs when provided.

This new added test creates a plot with twin axes where both axes have units. It then checks whether units are appended correctly on the respective axes. The code base without the modification fails the unit test whereas the modification makes it pass the unit test.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [n/a] New features are documented, with examples if plot related.
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [n/a] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

Fixes https://github.com/matplotlib/matplotlib/issues/22714

### CODE REVIEW

**Comment by weiji-li** [comment]
@oscargus What do I need to do in order to make this PR "workflow approved"?

**Comment by oscargus** [comment]
It is a default setting for new contributors. So once you have your first PR merged, all tests will run automatically for upcoming PRs.

**Comment by oscargus** [comment]
The label is just a reminder that this needs approval to run all tests. It doesn't affect anything else.

**Comment by weiji-li** [comment]
Thank you @oscargus ! Could you or any one review this PR?

File: lib/mpl_toolkits/tests/test_axes_grid1.py [hunk 1]
```diff
@@ -84,6 +85,43 @@ def test_twin_axes_empty_and_removed():
     plt.subplots_adjust(wspace=0.5, hspace=1)
 
 
+@image_comparison(['twin_axes_both_with_units.png'])
```
Comment (jklymak) [inline lib/mpl_toolkits/tests/test_axes_grid1.py:88]: You are changing the main library, but only testing the axes_grid toolkit.  If this has ramifications for the normal `Axes.twinx/y` please test that as well, or even preferentially.

**Comment by oscargus** [comment]
Sorry, I do not have enough knowledge about this module to be able to review it. (Just helping out with the formalities.)

File: lib/mpl_toolkits/tests/test_axes_grid1.py (hunk 1 above)
Comment (weiji-li) [reply lib/mpl_toolkits/tests/test_axes_grid1.py:88]: We have just added a new optional parameter to it, which should not affect the existing functionality. The regression tests for the main library have all been passed as well.

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py [hunk 2]
```diff
@@ -20,7 +21,8 @@ def __init__(self, parent_axes, aux_transform=None,
     def cla(self):
         super().cla()
         martist.setp(self.get_children(), visible=False)
-        self._get_lines = self._parent_axes._get_lines
+        self._get_lines = functools.partial(
```
Comment (jklymak) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py:24]: I don't understand this change - can you explain why you are doing this?

**Review by jklymak — COMMENTED** [review]
This had to be failing the main library as well, so this needs a test of units on twin axes in the main library.

**Comment by ksunden** [comment]
Took care of at least the rebase to keep up with changes in the library since this was opened.

I'm a little skeptical about the mechanism of using an undocumented argument to sneakily replace an instance attribute (at call time).

However, I do think I'm convinced that this is the most minimal change that addresses the problem at hand. It is at least only on an already private class.

Essentially the crux of the matter is that the `_process_plot_var_args` class manages a bit more of the state than perhaps it _should_ (but extricating that without a change to behavior is a much larger change)

The most concrete thing that is kept as part of this object is the color cycle state.
Simply removing the line modified by this PR actually fixes the units behavior, but at the cost of using a separate color cycle for the parasite axes. 
This is the behavior by `twin[x,y]` from standard axes, where the color cycle is separate but other behaviors tied to the axes such as units behavior work as expected.
If we were okay with that change in behavior, I think that would be my preferred fix for this issue, it has the benefit of being more consistent with twinx/y.

One alternative I explored, but think is fragile at best, was instead of replacing the `self._get_lines` object entirely, reaching in and taking the color cycle from the parent axes.
This actually works reasonably well in the short term, but would likely require quite a bit more machinery to behave as expected when `set_color_cycle` is called after the parasite axes are created.

The other direction that I think may be better, but haven't fully thought through is removing the `axes` reference from `_process_plot_var_args` entirely and shifting those calls into the base axes methods.

**Comment by anntzer** [comment]
@ksunden See also #19484 re: sharing of the color cycle state.  I don't think it's immediately usable in parasite_axes because right now the prop_cycle needs to be "declared" as shareable from the very beginning, but that's probably fixable, and using that in parasite_axes may be cleaner here.

**Comment by ksunden** [comment]
Moved to draft because I think this is not fully ready and the diff proposed above by @anntzer may prove to be a better route.

**Comment by melissawm** [comment]
Hello, @weiji-li - are you interested in moving this forward? Any chance you and @anntzer can collaborate on this?

**Comment by anntzer** [comment]
Superseded by https://github.com/matplotlib/matplotlib/pull/26078.  Thanks for the PR!

**Pull request closed without merging**